File: ts-claw/src/tools/read-file.ts
/** 读取文件工具。 */

import { readFile } from 'node:fs/promises';
import { join } from 'node:path';
import type { ToolDefinition } from '../schema/message.js';
import type { BaseTool } from './base.js';

export class ReadFileTool implements BaseTool {
  constructor(private workDir: string) {}

  name(): string {
    return 'read_file';
  }

  definition(): ToolDefinition {
    return {
      name: this.name(),
      description: '读取指定路径的文件内容。请提供相对工作区的路径。',
      inputSchema: {
        type: 'object',
        properties: {
          path: { type: 'string', description: '要读取的文件路径, 如 src/main.ts' },
        },
        required: ['path'],
      },
    };
  }

  async execute(arguments: string): Promise<string> {
    const args = JSON.parse(arguments) as { path: string };
    const filepath = join(this.workDir, args.path);

    const content = await readFile(filepath, 'utf-8');

    const MAX_LEN = 8000;
    if (content.length > MAX_LEN) {
      return `${content.slice(0, MAX_LEN)}\n\n...[内容过长，已被系统截断至前 ${MAX_LEN} 字节]...`;
    }

    return content;
  }
}

File: py-claw/src/py_claw/tools/read_file.py
"""读取文件工具。"""

from __future__ import annotations

import json
from pathlib import Path

from py_claw.schema.message import ToolDefinition
from py_claw.tools.base import BaseTool


class ReadFileTool(BaseTool):
    def __init__(self, work_dir: str) -> None:
        self.work_dir = Path(work_dir)

    def name(self) -> str:
        return "read_file"

    def definition(self) -> ToolDefinition:
        return ToolDefinition(
            name=self.name(),
            description="读取指定路径的文件内容。请提供相对工作区的路径。",
            input_schema={
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "要读取的文件路径, 如 src/main.py",
                    },
                },
                "required": ["path"],
            },
        )

    async def execute(self, arguments: str) -> str:
        args = json.loads(arguments)
        filepath = self.work_dir / args["path"]

        if not filepath.exists():
            raise FileNotFoundError(f"文件不存在: {filepath}")

        content = filepath.read_text(encoding="utf-8")

        MAX_LEN = 8000
        if len(content) > MAX_LEN:
            return f"{content[:MAX_LEN]}\n\n...[内容过长，已被系统截断至前 {MAX_LEN} 字节]..."

        return content
